src/DateTimeParser.jl

```diff
@@ -0,0 +1,512 @@
+                        res["hour"] += 12
+                    elseif res["hour"] == 12 && AMPM[lowercase(tokens[i+1])] == 1
+                        res["hour"] = 0
+                    end
+                    i += 2
+                else
+                    push!(ymd, round(Int, parse(Float64, token)))
+                    if i > len || !haskey(monthtovalue, lowercase(tokens[i]))
+                        i += 1
+                    end
+                end
+            elseif i <= len && haskey(AMPM, lowercase(tokens[i]))
+                # 12am
+                res["hour"] = round(Int, parse(Float64, token))
+                if res["hour"] < 12 && AMPM[lowercase(tokens[i])] == 2
+                    res["hour"] += 12
+                elseif res["hour"] == 12 && AMPM[lowercase(tokens[i])] == 1
+                    res["hour"] = 0
+                end
+                i += 1
+            elseif !fuzzy
+                error("Failed to parse date")
+            else
+                i += 1
+            end
+        else
+            # Token is not a number
+            if haskey(weekday, lowercase(token))
+                # Weekday
+                res["weekday"] = weekday[lowercase(token)]
+                i += 1
+            elseif haskey(monthtovalue, lowercase(token))
+                # Month name
+                push!(ymd, round(Int, monthtovalue[lowercase(token)]))
+                mstridx = length(ymd)
+                i += 1
+                if i <= len
+                    if tokens[i] in ("-", "/")
+                        # Jan-01[-99]
+                        sep = tokens[i]
+                        i += 1
+                        push!(ymd, round(Int, parse(Float64, tokens[i])))
+                        i += 1
+                        if i <= len && tokens[i] == sep
+                            # Jan-01-99
+                            i += 1
+                            push!(ymd, round(Int, parse(Float64, tokens[i])))
+                            i += 1
+                        end
+                    elseif (i+3 <= len && tokens[i] == tokens[i+2] == " " &&
+                            tokens[i+1] in PERTAIN)
+                        # Jan of 01
+                        # In this case, 01 is clearly year
+                        try
+                            # Make a guess
+                            value = parse(Int, tokens[i+3])
+                            # Convert it here to become unambiguous
+                            push!(ymd, convertyear(value))
+                        end
+                        i += 4
+                    end
+                end
+            elseif haskey(AMPM, lowercase(tokens[i]))
+                # am/pm
+                value = AMPM[lowercase(tokens[i])]
+                if value == 2 && res["hour"] < 12
+                    res["hour"] += 12
+                elseif value == 1 && res["hour"] == 12
+                    res["hour"] = 0
+                end
```

Comment:
You've repeated this logic several times. Think about abstracting it as a function